File: Day2/solution.ex
defmodule Solution do
  def input(filename) do
    File.read!(filename)
    |> String.split("\n", trim: true)
    |> Enum.map(&String.split/1)
  end

  @scores %{"X" => 1, "Y" => 2, "Z" => 3}

  def lose(opponent), do: Map.get(%{"A" => "Z", "B" => "X", "C" => "Y"}, opponent)
  def win(opponent), do: Map.get(%{"A" => "Y", "B" => "Z", "C" => "X"}, opponent)
  def draw(opponent), do: Map.get(%{"A" => "X", "B" => "Y", "C" => "Z"}, opponent)

  def round_score(opponent, self) do
    cond do
      rem(:binary.first(self) - :binary.first(opponent), 3) == 0 -> 6
      :binary.first(opponent) == :binary.first(self) - 23 -> 3
      true -> 0
    end
  end

  def part_one(data) do
    data
    |> Enum.map(fn [o, s] -> round_score(o, s) + Map.get(@scores, s) end)
    |> Enum.sum()
  end

  def part_two(data) do
    action = %{"X" => &lose/1, "Y" => &draw/1, "Z" => &win/1}

    score = fn [o, s] ->
      condition = apply(Map.get(action, s), [o])
      round_score(o, condition) + Map.get(@scores, condition)
    end

    data
    |> Enum.map(score)
    |> Enum.sum()
  end

  def main do
    data = input("input.txt")
    IO.puts "Part one: #{part_one(data)}"
    IO.puts "Part two: #{part_two(data)}"
  end
end

File: Day2/solution.kt
import java.io.File
import kotlin.math.max

enum class Weapon(val points: Int) {
    Rock(1),
    Paper(2),
    Scissor(3);

    companion object {
        fun fromInt(points: Int) = Weapon.values().first { it.points == points }
        fun fromKey(key: Char) = when (key) {
            'A', 'X' -> Rock
            'B', 'Y' -> Paper
            'C', 'Z' -> Scissor
            else -> null
        }
    }

    fun loses() = Weapon.fromInt(max(1, (points + 1) % 4))
    fun beats() = ((points - 1) % 4).takeUnless { it == 0 }?.let { Weapon.fromInt(it) } ?: Scissor
}

fun input(file: File) = file.readText().trim().split("\n").map {
    it.mapNotNull(Weapon::fromKey)
}

fun partOne(data: List<List<Weapon>>) = data.sumOf { (opponent, self) ->
    // Loss
    if (self.loses() == opponent) {
        self.points
    }
    // Draw
    else if (self == opponent) {
        self.points + 3
    }
    // Win
    else {
        self.points + 6
    }
}

fun partTwo(data: List<List<Weapon>>) = data.sumOf { (opponent, self) ->
    when (self) {
        // Lose
        Weapon.Rock -> opponent.beats().points
        // Draw
        Weapon.Paper -> opponent.points + 3
        // Win
        else -> opponent.loses().points + 6
    }
}

fun main() {
    val data = input(File("/home/vidar/Dev/AOC2022/Day2/input.txt"))
    println("Part one: ${partOne(data)}")
    println("Part two: ${partTwo(data)}")
}